Add monitoring alerts list and details pages

kyoto opened the pull request:
Adds an Alerts list page, an Alert details page and an Alert Rule details page.

All data is fetched from the Prometheus `/rules` endpoint (proxied through the backend).

The Alert and Alert Rule details pages use GET params to uniquely identify the object to be displayed because there is no unique ID associated with these objects.

Leaves the Prometheus and Alertmanager links in the sidebar for now (and renames "Alerts" to "AlertManager").

Left for future PRs:
- Labels list on the alert details page (because there is currently no Search page Alerts list to link to)
- End-to-end tests

<img width="1398" alt="screenshot-list" src="https://user-images.githubusercontent.com/460802/44198365-2bf08600-a17c-11e8-8d75-2f06ece9090a.png">

<img width="1399" alt="screenshot-alert" src="https://user-images.githubusercontent.com/460802/44198374-327efd80-a17c-11e8-9799-b7d3f528d66b.png">

<img width="1396" alt="screenshot-rule" src="https://user-images.githubusercontent.com/460802/44198383-3579ee00-a17c-11e8-86f3-65b6cd365f69.png">

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins push

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_411_build_416`

spadgett commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+
+let isLoading = false;
+const loadRules = () => {
+  const {prometheusBasePath} = (window as any).SERVER_FLAGS;
```
Should this be `prometheusBaseURL`?

kyoto replied:
Yes it should. Thanks!

kyoto commented on frontend/public/components/_alert.scss:
```diff
@@ -0,0 +1,25 @@
+$line-height: 18px;
```
Renamed to `$monitoring-line-height`

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+    return <span className="text-muted">{_.startCase(state)}</span>;
+  }
+  const klass = stateToIconClassName[state];
+  return klass ? <React.Fragment><i className={klass}></i> {_.startCase(state)}</React.Fragment> : null;
```
Done

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+        })
+        .catch(e => store.dispatch(UIActions.monitoringRulesErrored(e)));
+    };
+    setInterval(poller, 15 * 1000);
```
Thanks for pointing this out. I've changed it to be similar to that example and also fixed it to stop polling when the component unmounts.

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+export const AlertsDetailsPage = connectPage(connect(alertStateToProps)(AlertsDetailsPage_));
+
+// Convert duration in billionths of a second into a formatted string like "1h 30m"
+const formatDuration = (nanoseconds) => {
```
Thanks, I didn't notice that we already had a `formatDuration()` 😄 

The format is slightly different (doesn't include `d` for days and shows zero seconds like `5m 0s` instead of just `5m`), but I think it is OK.

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+      />
+      <StatusBox data={rule} loaded={loaded} loadError={loadError}>
+        <div className="co-m-pane__body">
+          <div style={{display: 'flex', justifyContent: 'space-between'}}>
```
Done

kyoto replied:
Done

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+                    <dd>{duration}</dd>
+                  </React.Fragment>}
+                  <dt>Expression</dt>
+                  <dd><pre style={{color: '#0088ce'}}>{_.get(rule, 'query')}</pre></dd>
```
Perfect 👍

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+      <NavTitle title="Monitoring Alerts" />
+      <div className="co-m-pane__filter-bar">
+        <div className="co-m-pane__filter-bar-group">
+          <span style={{marginRight: 20}}>OpenShift ships with a pre-configured and self-updating monitoring stack powered by <ExternalLink href={urls[MonitoringRoutes.Prometheus]} text="Prometheus" /></span>
```
Done

kyoto commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,450 @@
+    <div className="col-xs-6">Description</div>
+    <div className="col-xs-2">Active Since</div>
+    <div className="col-xs-2">State</div>
+    <div className="col-xs-2">Value</div>
```
Left this for now. Will discuss with @tlwu2013

spadgett replied:
FWIW we currently drop columns as the screen gets smaller, but we're looking at adding scrolling tables (#138)

alecmerdler commented on frontend/public/models/index.ts:
```diff
@@ -110,6 +110,28 @@ export const AlertmanagerModel: K8sKind = {
+  apiVersion: 'v1',
+  path: 'alerts',
+  abbr: 'AL',
+  plural: 'alerts',
```
`AlertModel` and `AlertRuleModel` need the property `crd: true`.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,427 @@
+  return {alert: null, name, rule: _.find(rules, {name, labels})};
+};
+
+class AlertsDetailsPage_ extends AlertsPageWrapper {
```
Skeptical of the need to use inheritance, which is usually discouraged in React because it muddies the lifecycle hooks. Prefer composition instead.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,427 @@
+  pending: 'fa fa-exclamation-triangle alert-pending',
+};
+
+const State = ({state}) => {
```
You can add `const State: React.SFC<StateProps>` and then define `StateProps` for better type safety.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,427 @@
+const connectToPrometheusURL = connectToURLs(MonitoringRoutes.Prometheus);
+
+const stateToProps = ({UI}) => UI.get(reduxID) || {};
+const connectPage = Page => withFallback(connect(stateToProps)(Page));
```
Nice use of `withFallback`!

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,427 @@
+  return {name, rule};
+};
+
+class AlertRulesDetailsPage_ extends AlertsPageWrapper {
```
Skeptical of the need to use inheritance, which is usually discouraged in React because it muddies the lifecycle hooks. Prefer composition instead.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,427 @@
+  return {filters: filtersMap ? filtersMap.toJS() : null};
+};
+
+class AlertsPage_ extends AlertsPageWrapper {
```
Skeptical of the need to use inheritance, which is usually discouraged in React because it muddies the lifecycle hooks. Prefer composition instead.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,427 @@
+}
+export const AlertsPage = connectPage(connectToPrometheusURL(connect(listStateToProps)(AlertsPage_)));
+
+interface Alert {
```
Prefer `type` to `interface`.

kyoto replied:
👍

alecmerdler commented on frontend/public/models/index.ts:
```diff
@@ -110,6 +110,28 @@ export const AlertmanagerModel: K8sKind = {
   propagationPolicy : 'Foreground',
 };
 
+export const AlertModel: K8sKind = {
```
This is not a native Kubernetes model, and therefore should not be included as `type K8sKind`.

kyoto replied:
Done and moved the models to `alert.tsx`. This means we can't use some of the util helpers like `ResourceIcon`, `resourcePath()` and `referenceForModel()`. I think it's better not to use them anyway since they are really just intended for k8s resources.

kyoto commented:
@alecmerdler Addressed all your comments as a separate commit (I'll squash before merging). Could you please take another look?

> Also don't know if defining models that overlap with the Prometheus Operator installed by OLM will cause problems.

As discussed offline, `GUI-Tests-OLM` were passing, so this should be fine.

alecmerdler commented on frontend/public/components/resource-pages.ts:
```diff
@@ -54,6 +55,8 @@ import {
 } from '../models';
 
 export const resourceDetailPages = ImmutableMap<GroupVersionKind | string, () => Promise<React.ComponentType<any>>>()
+  .set(AlertModelReference, () => import('./alert' /* webpackChunkName: "alert" */).then(m => m.AlertsDetailsPage))
```
Do not add components here because they are for Kubernetes resources only.

kyoto replied:
🎉

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,469 @@
+  return labels;
+};
+
+class AlertsPageWrapper extends SafetyFirst<AlertsPageWrapperProps, null> {
```
These components need unit tests.

kyoto replied:
I added a few unit tests, but I'd like to leave the e2e tests for a follow up PR.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,469 @@
+type AlertsPageWrapperProps = {
+  Page: React.ComponentType<any>;
+};
+type AlertsDetailsPageProps = {
```
Export theses types so they can be used in unit tests.

kyoto replied:
👍

alecmerdler commented on frontend/public/components/app.jsx:
```diff
@@ -166,6 +167,10 @@ class App extends React.PureComponent {
           <LazyRoute path="/k8s/cluster/clusterrolebindings/:name/copy" exact kind="ClusterRoleBinding" loader={() => import('./RBAC' /* webpackChunkName: "rbac" */).then(m => m.CopyRoleBinding)} />
           <LazyRoute path="/k8s/cluster/clusterrolebindings/:name/edit" exact kind="ClusterRoleBinding" loader={() => import('./RBAC' /* webpackChunkName: "rbac" */).then(m => m.EditRoleBinding)} />
 
+          <Route path="/k8s/cluster/alerts" exact component={AlertsPage} />
```
The `/k8s/` route prefix is reserved for Kubernetes resources only. Put these routes under a different prefix since they are not Kubernetes resources.

kyoto replied:
Changed the route prefix from `/k8s/cluster/` to `/monitoring/`.

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,469 @@
+import { BreadCrumbs, history, NavTitle, SectionHeading, StatusBox, Timestamp, withFallback } from './utils';
+import { formatDuration } from './utils/datetime';
+
+const AlertModel = {
```
Just name this `Alert` to avoid confusion with existing concept of "models".

kyoto replied:
👍

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,469 @@
+};
+export const AlertModelReference = `monitoring:v1:${AlertModel.kind}`;
+
+const AlertRuleModel = {
```
Just name this `AlertRule` to avoid confusion with existing concept of "models".

kyoto replied:
👍

alecmerdler commented on frontend/public/components/alert.tsx:
```diff
@@ -0,0 +1,469 @@
+  path: 'alerts',
+  abbr: 'AL',
+};
+export const AlertModelReference = `monitoring:v1:${AlertModel.kind}`;
```
"References" are the `GroupVersionKind` representation of a Kubernetes resource, which neither `Alert` nor `AlertRule` are.

kyoto replied:
Removed

kyoto commented:
@alecmerdler I added another commit to address your latest feedback. Could you please take another look?

alecmerdler commented:
jenkins rebuild

kyoto commented:
@alecmerdler @spadgett 

I rebased and squashed the commits and also switched the `LazyRoute`s back to `Route`s since they were not working properly and we will leave lazy loading for a follow up.

Can I get a `/lgtm`? 😄
